Refactor Abins, replace SData with Euphonic data structures

ajjackson opened the pull request:
### Description of work

#### Summary of work
Remove `abins.sdata.SData`, use classes from `euphonic.spectra` instead

#### Purpose of work
Abins currently manages collections of simulated spectra with an "SData" class. This assumes that the data will be structured in a tree with heirarchy atom -> quantum-order. This is not flexible enough to deal with coherent scattering, which might not be decomposable by atom.

Euphonic is an existing dependency and uses Spectrum1D, Spectrum2D and Spectrum1DCollection classes to manage spectra including arbitrary metadata. Effectively this is a flat table structure, amenable to "query" or "pipeline" workflows. This seems to be an appropriate tool, so here we extend that to add the missing Spectrum2DCollection and work towards using those natively in Abins.

#### Further detail of work
- [x] Add a method to SData to produce an equivalent SpectrumNDCollection object
- [x] Update abins to use these objects for all "output" management (i.e. summing relevant subsets and creating workspaces)
- [x] Update abins to simulate data directly into a SpectrumNDCollection object
- [x] Delete SData class and related unit tests
- [x] Move patches and extensions to Euphonic classes upstream into a new euphonic release
- [x] Remove Euphonic subclasses from mantid, update Euphonic dependency version

### To test:

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

ajjackson committed:
Extend broadening efficiency tweak to 2D

Not seeing a big impact on test timings here, but this _should_ scale
better as we add more atoms.

ajjackson committed:
Format mass metadata to string

Officially the Spectrum metadata should not include float because they
can't be reliably compared for equality

ajjackson committed:
calculate_s_isotropic directly into Spectra

For now, fallback via SData if no isotroptic calculation

ajjackson committed:
Continue pushing SData further back into nested function heirarchy

ajjackson committed:
Start saving structure factors directly to Spectrum objects

This is a significant step towards removing SData entirely.

For compatibility at this point, Spectrum objects mostly store the bin
centres rather than bin edges. When SData is gone it would nicer to
move to storing bin edges. (Centres can be computed from edges, but
edges can only be guessed from centres...)

ajjackson committed:
Change return type of spowdersemiempiricalcalculator

- Now the Algorithm classes never deal with SData at all
- SData still used internally and cached

ajjackson committed:
Implement _atom_type_s using Spectrum1DCollection features

- Although Spectrum1DCollection provides a .select() function, we
  ended up using a chain of filters here because it can't handle
  numerical comparisons. Perhaps this would be a bit cleaner if
  the select() method accepted our filter functions.
- The Spectrum1DCollection group_by() method throws an error if the
  key is in the top level of metadata (i.e. because it applies to all
  rows.) It would be good to clean that upstream too.

ajjackson committed:
More spowdersemiempiricalcalculator refactoring

- Push SData a bit deeper in the chain of functions

- Stop using SDataByAngle, instead use angle metadata in
  SpectrumCollection objects

- Patch SpectrumCollection so that group_by() and select() work as
  expected (i.e. run succesfully) when some keys are at top level of
  metadata (i.e. common to all lines).

- Apply broadening to SpectrumCollection objects. For now we keep
  using Abins implementation of broadening so that tests pass; this
  should be replaced with Euphonic implementation later.

ajjackson committed:
Try to make Spectrum Collection concatenation more efficient

The .from_spectra() constructor has a lot of overhead as it checks the
types and numerical agreement of axes; if we are comparing Collections
this should only be checked once.

If this tweak works well it should move upstream to euphonic. Here we
refer to a private method on the 1D collection from the 2D collection:
really this can live on a parent class.

ajjackson committed:
Slightly improve performance of spectrum broadening

Remove innermost loop of Spectrum creation by assembling y_data array
directly.

ajjackson committed:
More Abins spectra->workspaces tidy-up

ajjackson committed:
Drop SData autoconvolution method, port test to replacement function

ajjackson committed:
Apply Debye-Waller factor to SpectrumCollection, drop SData method

Continue dissolving SData: apply DW to output spectra, drop now-unused
method from class

ajjackson committed:
Refactor threshold check so it can be done outside SData

Move the threshold-check logic to module-level functions, and leave
existing method as a wrapper. Then we should be able to delete the
class later...

Also remove some useless parameters/logic that make the return value
optional. We can just ignore the return value if we don't care about
it.

ajjackson committed:
Remove Abins SData class

Finally, the cathartic part of the refactor where I get to delete 500
lines of code at once

ajjackson committed:
SData refactor: allow Spectrum1DCollection to be produced

To begin migration from SData to euphonic data types, allow existing
SData to produce a Spectrum1DCollection. Then we can start
modifying code that _uses_ SData to use this datatype instead.

So far we modify the _atom_number_s method which creates output
workspaces from the Abins calculation.

Spectrum2DCollection is not implemented yet, so initially we leave the
old implementation in place for 2D cases.

ajjackson committed:
Minor cleanup

ajjackson committed:
Extend Euphonic Spectrum types in abins.sdata

This is not very thoroughly tested here, and should be moved upstream
once everything looks happy on Abins side.

- Wrap Spectrum1DCollection to deal with group_by() issue where
  metadata is the same for all items (and so not in line_data)
- Implement Spectrum2DCollection
- Have SData.get_spectrum_collection() dispatch to 1D or 2D as
  appropriate
- Tweak workspace writers to use either SpectrumNCollection class

ajjackson committed:
Remove unused SDataByAngle class and tests

The Spectrum1DCollection/Spectrum2DCollection classes allow arbitrary
metadata, so there is no need for a technical distinction between the
class that collects over atoms and quantum orders and one that
collects over angles.

ajjackson committed:
Tidy up import

ajjackson committed:
Return suitable isotropic data quickly, drop empty SData builder

- Rather than go through the data transformation on arrays of zeros,
  we can return an almost-empty SpectrumCollection immediately.
- That finally eliminates SData from this module! There is definitely
  room for optimisation as autoconvolution has become more expensive,
  but the order-1 stuff seems to be running in similar time. It would
  be a good idea to do some profiling later...

- As part of rebasing this, restored some missing units to
  spowdersemiempiricalcalculator and fixed AbinsSDataTest. Sorry, it
  should be on another commit but this is a big job and would benefit
  from some squashing later.

ajjackson committed:
Rework abins "masses table" to work with spectrum

As the relevant information is available in spectra, this allows us to
drop a redundant parameter and reduce the number of objects passed around

ajjackson committed:
Move threshold check, fix 2D systemtest

Bin checks on Spectrum2DCollection need adapting to interpret 3-D
z_data array correctly. This implementation is a bit inefficient as it
needlessly instantiates a Spectrum2D; leave more verbose/efficient
implementation for upstream.

ajjackson committed:
Correct axis-order for dummy 2D spectra

This was breaking a doctest

ajjackson committed:
Factor out kinematic constraints from SData

This functionality is also available in Euphonic, and it would make
sense to remove this entirely from Abins and use it there.

However
- this will need a bit of tinkering to work with Spectrum2DCollection,
  and doing it upstream would avoid a lot of duplication
- it would be better to apply kinematic constraints (and instrument
  broadening) after spectra have been filtered and summed into output
  groups

So let's defer that for now

ajjackson committed:
Apply autoconvolution, q2 corrections to spectra

- create new autoconvolution routine that works with spectrum
  collection

- integrate downsampling into autoconvolution loop over atoms

  - this could save a bit of memory, or benefit from parallism over
    atoms

  - not clear if benefit is large at this point, but it doesn't add
    much complexity either. YAGNI?

- apply q2 order corrections, converting 1D initial spectra to 2D maps
  for quantum orders with isotropic Debye-Waller correction (i.e. all
  orders above fundamental)

- drop _broaden_sdata as this is no longer needed

ajjackson committed:
Tidy up comments and references to SData

ajjackson committed:
Handle fundamental-only cases more robustly

- The euphonic SpectumCollection select() method doesn't work when all
  lines in the collection match the criterion, because such metadata is
  moved into the top-level metadata rather than the per-line section
  which is checked. The key is therefore treated as missing.

- Compounding this, the data container cannot be entirely
  empty, so it allows an error to be raised if no matches are found.

- More reasonably, the `from_spectra()` constructor cannot operate on
  an empty list, as this means no (common) axis information woudl be
  available. In this commit we avoid calling this when no data is available.

ajjackson committed:
Further optimisation of autoconvolution: Dealing with Pint overhead

Spectrum.x_data (etc.) properties are quite expensive to call as they
trigger a call to ureg to build Units and call Quantity.to() to ensure
output unit is consistent with the stored unit name.

Try to avoid calling it inside a loop! In practice this means using
private attributes a bit more; not ideal, but we pin the Euphonic
version and I maintain the package so for now it is safe.

Avoid calling `* ureg` where we know we have an array and can
build Quantity directly. This saves on some type-checking.

Implement an "unsafe" constructor for Spectrum1DCollection from a
series of Spectrum1D when we already know that the axes will be
consistent. Again this saves a lot of time spent on consistency checks.

ajjackson committed:
Implement iter_metadata for spectrum collections, speed up loops

This iterator makes it more practical to iterate over the content of a
spectrum collection without (slowly) creating new Spectrum objects

Profiling shows a dramatic impact in the places changed so far.

ajjackson committed:
Parallelise Abins autoconvolution over atoms

As the new data container facilitates a more "functional" programming
style with generators, it is quite simple to incorporate
multiprocessing and get the best performance we can from a typical
IDAaaS instance.

Test calculations are showing that with a few cores we get a factor of
2x or so, which seems worth having on what can be a bottleneck
step (if there are a lot of atoms and few k-points, which is typical
for molecular calcualtions.)

I expect that this step will be decoupled from other possible
parallelism over k-points or atom intensity calculations, so it can
use the full available processor count.

ajjackson committed:
More efficient spectrum slicing

Instead of deconstructing to individual spectra, filtering and
reconstructing, just look at the metadata and then index the
collection in one go. This is a lot more efficient as it avoids much
copying of x_data etc.

The [spectrum for spectrum in ...] line was previously taking 70% of
the function run time in a test calculation.
Now it is dominated by _calculate_s_isotropic (97% of time) as it
should be.

ajjackson committed:
Implement parallel broadening over 1D spectra

A simple for-loop over data rows is converted to a Pool.starmap for
an easy speedup.

For a benzene test case on a 30-core EPYC VM, this reduced time in the
function from 1.85s to 0.43s

ajjackson committed:
Tidy up: for now 2D parallel broadening is not implemented

ajjackson commented:
This initially led to some severe performance regressions, due to the relatively high overhead of iterating over a Euphonic Spectrum1DCollection/Spectrum2DCollection object, which performs unit conversions while spawning new objects. Some efficiency tweaks can be made to the Euphonic objects, but mostly we can recover the performance by iterating in other ways (over the metadata and underlying arrays).

In addition, we can start to introduce `multiprocessing`-based parallelism, which plays nicely with loops over "flat" tables.

ajjackson committed:
Abins2D optimisation: avoid unnecessary Spectrum iteration

Iteration over metadata is much cheaper than spawning new spectrum
objects.

ajjackson committed:
Experiment with parallel 2D broadening

With my test case of a Benzene cell (one k-point, fair number of
atoms, 10th-order expansion): by default, parallelising the inner loop
over 2D spectra gives a decent speedup as Pool is increased to around
15-16 cores, then deterioriates as we work up to the 30 cores
available.

This seems is significantly improved by setting a chunksize of 5, but we
leave this adjustable in abins.parameters. (There is still a *slight*
edge at some optimal core count, but it is no longer worth the user
effort of checking!)

ajjackson committed:
Better 2D broadening: remove an outer loop

By converting the data to a 2-D array and reshaping at the end, we get
more data into a single starmap() call and better parallel
efficiency. At this point it is also helpful to increase the chunksize
a bit: at 100 the cost increase to 1-D is slight and 2-D benefits a
lot.

Perhaps they should be different parameters and more
benchmarking/adaptive setting used to optimise. But for now this is a
lot better than serial operation!

Performance is now satisfactory, I will not attempt a shared-memory
implementation for the current branch (which is supposed to be focused
on data refactor...)

ajjackson committed:
Tidy up an import statement

ajjackson commented:
Very strange test failure there: something is going wrong with the Pychop call inside a multiprocessing starmap, but only for OSX and Windows. I wonder if there is some OS-dependent namespace hackery that runs into issues because both abins and Pychop have an Instrument class 🤔

ajjackson committed:
Cleaner parallel map with Pool.map and functools.partial

On previous attempt I thought this wouldn't work and that starmap was
required.

The partial function seems a cleaner way of making it clear that only
one parameter is actually being mapped over. It does also seem to work
with a staticmethod, so we can bring the function definition closer to
where it is used.

ajjackson committed:
TMP: try converting parallel map of broadening to series

Getting strange pychop CI errors on Windows and Mac. See if it is
specific to multiprocessing...

ajjackson committed:
Precompute Pychop resolution before parallel map on spectra

Mac/Win tests are failing when the Pychop resolution is computed in
parallel. We don't really need to do that on every thread though; we
can compute the energy function once before distributing.

ajjackson committed:
Create "prepare_resolution()" instrument method

This moves responsibility for method-specific details to the
instruments; at a higher-level we can just call this method to ensure
that we are in a good position for parallel computation.

sf1919 commented:
Apologies... wrong PR

sf1919 commented:
@ajjackson is this PR still one to be worked on?

ajjackson commented:
Yes it is still in progress. Effort moved upstream to Euphonic for a bit and will be back here shortly (with many deleted lines of code.)

ajjackson committed:
bump euphonic requirement

ajjackson commented:
Oof, that was a big rebase

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ajjackson committed:
Move type-hint definition to before first use

ajjackson committed:
Revert "bump euphonic requirement"

This reverts commit 75f77f1b214330734697776929e377a264d2af61.

Let's see if the CI will pass before doing this version bump and
cleaning up; currently some code depends on private attributes that
have changed in Euphonic 1.4.

ajjackson committed:
Pin to pre-refactor Euphonic while checking tests on rebase

ajjackson committed:
Bump euphonic requirement

- 1.4.2 fixes an issue with importing base Spectrum

ajjackson committed:
(mostly) delegate Spectrum Collection implementations to Euphonic

I found one bug which we'll patch here for now...

ajjackson commented:
Hurrah, we now have a substantial net reduction in lines-of-code 🎉

ajjackson committed:
Add missing get_bin_edges method to Spectrum2DCollection

This should be moved upstream but it's helpful to start here and check
on the Mantid CI...

ajjackson committed:
Bump Euphonic, drop patched code

ajjackson commented:
I have done a few large benchmark calculations now:
- for this paper we ran Abins on supercells of ~1500 atoms https://arxiv.org/abs/2504.19352 ; this branch and `main` each took around 90 mins for calculations
- in a separate benchmark using some NaCl force constants from the Euphonic training set (i.e. a small unit cell) with a large q-space sampling  of ~97,000 q-points, both branches took 6.4 min to run

So I am confident that there is not a significant negative performance impact. (There _is_ certainly scope to improve the performance in those cases by better use of parallelism, though!)

ajjackson committed:
Remove wrappers for Euphonic spectrum classes

ajjackson committed:
Relax Euphonic upper version bound

ajjackson commented:
I'm sorry that this diff is a bit tricky to navigate. The github view does rather too much interweaving of blocks! I'll try to break down what is going on file-by-file to help with review.

At a high level: Abins calculates many individual contributions to the inelastic phonon scattering, by atom and by quantum order. These have been stored in a class "SData" which is essentially a nested dict. However the design of the dict was insufficiently flexible to support future functionality: it assumes that we can always organise the contributions by Atom and by individual quantum orders, and that no other divisions will be worth keeping track of. Already an "SDataByAngle" collection was added to the scheme, complicating things.

Here we replace that with data structures that are already maintained and tested in Euphonic: Spectrum1DCollection and Spectrum2DCollection, which are collections of Spectrum1D and Spectrum2D. These are "flat" collections which allow contributions to be managed with arbitrary metadata of the form `dict[str, str|int]`. There is no longer a special organisation for angles: this is just another bit of metadata in the "table". But there is a stricter separation of 1D vs 2D into separate classes: SData uses the same class for either an one must figure out which is being used from the data shape or presence of q bins.

Unlike Abins, which works in Angstrom and wavenumber internally, Euphonic uses the package Pint to attach units to data and facilitate conversion between them. So wherever we see a Quantity, this represents a scalar or numpy array with associated unit. They can be created with the Quantity() constructor or by multiplying with a unit produced by `ureg`.

The most significant changes in the PR are:
- Entirely remove the existing SData class from *sdata.py*
- Create/rework a few functions in *sdata.py* that provide Abins-specific functionality to the Euphonic classes
- Rework the various iterations and construction of data in *spowdersemiempiricalcalculator.py* (the actual intensity calculation), *abinsalgorithm.py* (a common base class to Abins and Abins2D algorithms), and the *Abins.py*, *Abins2D.py* and *abins2.py* algorithm implementations.
- By flattening the iteration it became simpler to implement parallelism; this is implemented for broadening and autoconvolution stages.

The good news is that (I think) the new iteration schemes look a bit nicer.

File-by-file notes:
- Abins.py/Abins2D.py/abins2.py
  - `sdata` is renamed to `spectra` to allude to underlying type
- **Abins2BasicTest.py**
  - Restrict the number of quantum orders in test calculations; these were running slowly and it was inconsistent with the Abins-v1 tests
- conda_build_config.yaml
  - Bump the required Euphonic version to get support for Spectrum2DCollection and improvements to all spectra
- abinsalgorithm.py
  - **clean up some logic for early exit**
  - Rework get_masses_table to use new data structure
  - Rework create_workspaces to use new data structure
- **instrument.py, pychop.py**
  - factor out Pychop-fitting step so that it can be triggered from outside the `calculate_sigma` step. This is to support the parallel broadening implementation.
- parameters.py
  - the existing `threads=4` parameter was not actually used anywhere, and is not a particularly good default. Instead, define a chunk size and detect available CPUs automatically.
- sdata.py
  - remove SData, SDataByAngle
  - replace missing functionality with functions operating on Spectrum objects
- spowdersemiempiricalcalculator
  - use spectrum classes instead of SData
  - simplify case where fewer quantum orders are requested than previous calculation: just load all of them! When workspaces are constructed we will only build/fill them for the requested orders.
  - implement parallel broadening with multiprocessing

The test suite has been changed as little as possible to be confident that the results are unaffected. This means that the reference data is still in a format related to the old data types, and converted on the fly to something that can be compared more directly. When the dust has settled, the tests could be cleaned up a bit by converting the relevant data to Euphonic JSON formats.

In **bold** I have highlighted a few of the changes that are not strongly coupled to the spectrum work. If it helps I could cherry-pick these into a separate PR which can be quickly reviewed and slightly reduce the main diff?

sf1919 commented:
The failing test looks suspicisouly similar to the pydantic failure - #39131

ajjackson commented:
The new pin was just merged so I'll try a rebase.

ajjackson commented:
Here is a PR with the self-contained changes, if that helps tidy up the diff! https://github.com/mantidproject/mantid/pull/39343

adriazalvarez reviewed (changes requested):
Thanks for this change.  It seems the euphonic data structures blend very well with abins, as stripping most of `sdata` makes the code base more easy to read. 
All tests are passing, I have only a few small comments.

adriazalvarez commented:
Other comment, not related to this PR. I think the euphonic docs don't have an entry for `Spectrum2DCollection`, it's easy to guess what it does, but maybe for completion you may consider adding it at some point.

ajjackson commented on scripts/abins/abinsalgorithm.py:
```diff
@@ -408,61 +410,70 @@ def get_atom_selection(*, atoms_data: abins.AtomsData, selection: list) -> Tuple
     @staticmethod
-    def get_masses_table(atoms_data):
+    def _iter_line_metadata(metadata: Dict[str, int | str | List[Dict[str, int | str]]]):
+        """Iterate over spectrum metadata as though all data were in line_data"""
```
Will take a look, it may be that an equivalent function exists in euphonic now anyway.

This part:
```python
            for key in metadata.keys()
            - {
                "line_data",
            }
```

is an example of `black` formatting obscuring the meaning!

`for key in (metadata.keys() - {"line_data",})`

makes it clearer what we are really iterating over: a difference of sets.

ajjackson committed:
Use existing euphonic method to iterate over metadata

Spectrum collection store metadata as a mixture of common attributes
and line-specific, e.g.

{"element": "Na",
 "scattering": "Coherent",
 "line_data": [{"atom_index": 1}, {"atom_index": 2}]
}

The constructors and iteration methods on these classes split and
merge as needed. When we want to only iterate over the metadata, it is
wasteful to iterate over the whole collection (i.e. spawning new
Spectrum objects) so we use the iter_metadata() method which doesn't
touch the main data arrays.

This wasn't always available in euphonic so was initially duplicated here.

ajjackson committed:
Quick fixes from code review

- Correct docstring for check_thresholds()

- Avoid unnecessary private access to spectra._z_data when masking
  kinematically-inaccessible values.

  Spectrum objects use a ``@property`` abstraction to store raw arrays
  and units rather than Quantity. This is intended to benefit the
  compiled C extensions by providing a fast private interface to numpy
  arrays that doesn't have to deal with Pint.

  I'm not sure at what point the "set" interface started working
  nicely again, but the classes were recently refactored to use Pint
  more efficiently and a bug may have been fixed in the process.

ajjackson committed:
Rework autoconvolution to use Spectrum1DCollection

This method was working with lists of spectra in order to avoid some
overhead in constructing and iterating over the sequences. Between an
upstream refactor of the .y_data method and the addition of an "usafe"
option to .from_spectra, this should no longer be necessary.

This allows us to work reasonably efficiently without using private
attributes.

ajjackson committed:
Raise error if max autoconvolution order is too low (i.e. = 1)

This is not a likely scenario but would lead to some surprising
behaviour, so let's raise an error.

ajjackson committed:
Review feedback: stricter cast float to string

16 decimal places should be enough to distinguish any Python float

ajjackson committed:
Expand docstring

ajjackson committed:
Review feedback: 1D/2D handling in spowdersemiempiricalcalculator

Clean up these checks by setting a single flag at the level of the
calculator class.

Implemented as a cached property to keep API clean and avoid repeated
lookups from instrument list.

Perhaps the lists could be dispensed with by storing this information
on the instruments themselves. But we would still need a list of them
_somewhere_ for the factory function.

ajjackson commented on scripts/abins/spowdersemiempiricalcalculator.py:
```diff
@@ -479,75 +565,78 @@ def _isotropic_dw(*, q2, a_trace):
+        if isinstance(spectra, Spectrum1DCollection):
+            frequencies = spectra.x_data.to(self.freq_unit).magnitude
+
+            with Pool(n_threads) as p:
```
What scenario do you have in mind? A timeout in case of process communication error? I don't know how reliably we can predict how long this operation should take, maybe with a bit of testing that can be done.

ajjackson commented on scripts/abins/spowdersemiempiricalcalculator.py:
```diff
@@ -326,75 +354,134 @@ def _calculate_s_isotropic(self, q2: np.ndarray, broaden: bool = False) -> SData
-
         # Get numpy broadcasting right: we need to convert an array from (order,)
         # to (order, energy) or (order, q, energy) format.
         if len(q2.shape) == 1:
```
Good idea!

ajjackson committed:
Use already-defined units

ajjackson commented on scripts/abins/spowdersemiempiricalcalculator.py:
```diff
@@ -617,11 +716,43 @@ def _calculate_fundamentals_over_k(self, angle: float = None, q2: np.ndarray = N
+        match q2.shape:
+            case (n_q, 1):
+                return Spectrum2DCollection(
+                    x_data=(self._q_bins * ureg("1/angstrom")),
```
Good catch! For now it doesn't make much difference as this is the only implementation in Abins that matters. It could make sense to centralise in abins.constants later. I'd be reluctant to define them in abins.parameters as this invites more fiddling and it's not guaranteed all the conversions will be done correctly. (In Euphonic they would be...)

ajjackson commented:
> Other comment, not related to this PR. I think the euphonic docs don't have an entry for `Spectrum2DCollection`, it's easy to guess what it does, but maybe for completion you may consider adding it at some point.

Cheers, opening a Euphonic Issue for this https://github.com/pace-neutrons/Euphonic/issues/413

adriazalvarez reviewed (dismissed):
Thanks for the changes.

ajjackson commented:
Thanks for review, hope it wasn't too painful relative to size!

cailafinn reviewed (changes requested):
Looking very good, just have a couple of questions before merge.

ajjackson commented on scripts/abins/abinsalgorithm.py:
```diff
@@ -507,93 +501,95 @@ def _create_workspace(self, *, species: AtomInfo, s_points: np.ndarray, label: s
-        Helper function for calculating S for the given atomic index
+    def get_s(self, spectrum: Spectrum) -> Quantity:
+        """Get spectral quantity array, from y or z axis of Spectrum as appropriate"""
+        from abins.constants import ONE_DIMENSIONAL_INSTRUMENTS
```
No technical reason, just subjective tidiness which may have changed between iterations of refactoring! I could move it to the top if you like.

ajjackson commented on scripts/test/Abins/AbinsSDataTest.py:
```diff
@@ -4,468 +4,67 @@
-        self.assertTrue(np.allclose(input_data.get_frequencies(), output_data.get_frequencies()))
-        self.assertTrue(np.allclose(input_data.extract()["atom_0"]["s"]["order_1"], output_data.extract()["atom_0"]["s"]["order_1"]))
-
-    def test_sdba_from_sdata_series_bad_inputs(self):
```
They test classes which no longer exist in Mantid. The functionality was replaced by classes from Euphonic, which has 97% code coverage. (That doesn't mean things can't slip through, but it's a good start!)

cailafinn replied:
Great, thanks!

ajjackson committed:
Move abins.constants imports to top of file

ajjackson committed:
Simplify abins.parameters.performance["threads"] check

- Use stdlib multiprocessing, which is actually used by the code
- import at top level
- accept None gracefully (the default and recommended setting)

ajjackson committed:
Move another import to top-level

cailafinn reviewed (commented):
Great, looks good to me now. Just wanted to check if this has any user-facing behaviour changes? If so, could we have a release note?

ajjackson commented:
It's not supposed to, unless they are importing things from the scripts directory and fiddling with advanced parameters. The performance profile will have slightly changed but I benchmarked this and it doesn't seem to be noticeable at this point.

Perhaps it is worth writing that there was a big refactor of the internal library for AbINS, as it could _in theory_ impact users who do clever things with external scripts that hook into AbINS. Most likely, I am the only one of those 😅
